docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,366 @@
+```
+
+Alias is a token that will be replaced with the image reference in the builder definition files. If no value is provided for `alias` then it defaults to the value of `image`.
+
+## Config validation
+
+We add three new validations to the [validation](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/schema/validation/validation.go#L37) package after the introduction of artifact dependencies:
+- Cyclic references among artifacts.
+  - We cannot have image `A` depend on image `B` depend on image `C` depend on image `A`.
+  - We run a simple depth first search cycle detection algorithm treating our `Artifact` slice like a directed graph- image `A` depending on image `B` implies a directed edge from `A` to `B`.
+- Unique artifact aliases.
+  - We ensure that within *each* artifact dependency slice the aliases are unique. 
+- Valid aliases
+  - We validate that `alias`es match the regex `[a-zA-Z_][a-zA-Z0-9_]*` for `ArtifactDependency` defined in `docker` and `custom` builders since these are used as build args and environment variables respectively.
+
+## Referencing dependencies
+
+### Docker builder
+
+The `docker` builder will use the `alias` of an `ArtifactDependency` as a build argument key.
+ 
+ ```yaml
+ build:
+  artifacts:
+    - image: simple-go-app
+    - image: leeroy-app
+      requires:
+        - image: simple-go-app
+          alias: BASE
+ ```
+
+Here the alias is used to populate a build arg `BASE=gcr.io/X/simple-go-app:<tag>@sha:<sha>` passed along with a `--build-arg` flag (or a buildKit parameter) when building `leeroy-app`.
+
+### Custom builder
+
+The `custom` builder will be supplied each `ArtifactDependency`'s `alias` and image reference as environment variables keyed on `alias`. So they can be easily referenced in user-defined build definitions.
+
+### Buildpacks builder
+
+Buildpacks supports overriding the run-image and the builder-image in its current schema. We extend this to allow specifying `ArtifactDependency`'s `image` name as the value for the `runImage` and `builder` fields.
+
+```yaml
+build:
+  artifacts:
+  - image: builder-image
+  - image: run-image
+  - image: skaffold-buildpacks
+    buildpacks:
+      builder: builder-image
+      runImage: run-image
+    requires:
+      - image: builder-image
+      - image: run-image
+```
+
+If there are any additional images in the `required` section it only enforces that they get built prior to the current image. However, the buildpacks builder cannot really reference them in any other way.
+
+### Jib builder
+
+The Jib builder supports [changing the base image](https://cloud.google.com/java/getting-started/jib#base-image). We add a new field `baseImage` to the builder definition that can be set to an `ArtifactDependency`'s `image` field.
+
+For Maven:
+
+```yaml
+build:
+  artifacts:
+  - image: base-image
+  - image: test-jib-maven
+    jib:
+      type: maven
+      baseImage: base-image
```

Comment:
We have some new code that allows the user to specify the target platform that might conflict with some of the ideas here (for instance if the original image is a manifest list and the user is selecting a specific platform). However, I don't think that's necessarily within scope of this doc and I'm not even sure it will come up for a 99% of users. Just wanted to make a record of it in case we need to handle it later.

Reply:
thanks for that info. We can take that up as a follow up work if it ever comes to it (since 99% of users should be unaffected)